### PR TITLE
RFC: Pickle Protocol for Keras

### PR DESCRIPTION
**Pull request by adriangb**
Comment period is open until 10-13.

As per discussion in https://github.com/tensorflow/tensorflow/pull/39609#issuecomment-685869003

| Status        | Proposed       |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [286](https://github.com/tensorflow/community/pull/286) |
| **Author(s)** | Adrian Garcia Badaracco ({firstname}at{firstname}gb.com), Scott Sievert (tf-rfc@stsievert.com)  |
| **Sponsor**   | Mihai Maruseac (mihaimaruseac@google.com)                 |
| **Updated**   | 2020-09-21                                           |

## Objective

Implement support for Pickle, Python's serialization protocol within Keras.

### CODE REVIEW

**Comment by adriangb** [comment]
@stsievert

**Comment by adriangb** [comment]
I think one important thing to not is that ideally we want to use `SaveModel`/`Model.save` since that saves optimizer states and other things like that. The limitation is that that entire ecosystem is hardcoded to only work with filesystem files. I traced that back to [here](https://github.com/tensorflow/tensorflow/blob/b36436b087bd8e8701ef51718179037cccdfc26e/tensorflow/python/lib/io/file_io.py#L333).

**Comment by adriangb** [comment]
cc @mihaimaruseac for initial review

**Comment by mihaimaruseac** [comment]
It seems there are no more comments. Let's schedule an internal review then. I'll come back with the date.

**Comment by mihaimaruseac** [comment]
Design review scheduled for 22nd of October at 10am Pacific Time.

**Comment by adriangb** [comment]
> Design review scheduled for 22nd of October at 10am Pacific Time.

Awesome, thank you!

File: rfcs/20200921-pickle-for-keras.md [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+  Pickle; however, many parallelization libraries don't
+  special case Keras models (e.g, Joblib). Relevant errors are most common in hyperparameter
+  optimization with Scikit-Learn's parallelization through Joblib
+  ([TF#33204] and [TF#34697]) or parallelization through Dask ([dask-ml#534]).
+* Lack of Pickle support can complicate saving training history like in
+  (the poorly asked) [SO#54070845].
+
+This RFC would resolve these issues.
+
+## Design Proposal
```
Comment (fchollet) [inline rfcs/20200921-pickle-for-keras.md:119]: What is the list of base classes that you expect will need to implement the `__reduce__` method?

Comment (adriangb) [reply rfcs/20200921-pickle-for-keras.md:119]: I think a good starting point would be:
- `tf.keras.Model`
- `tf.keras.metrics.Metric`
- `tf.keras.losses.Loss`
- `tf.keras.optimizers.Optimizer`

I think layers are usually serialized as part of a `Model`; I don't think there's much of a use case for implementing `__reduce__` for individual layers.

But generally, all classes that already have a `serialize` and `deserialize` method could benefit getting a `__reduce__` that just references those, but there is no reason why they have all be implemented together (or at all).

**Comment by ematejska** [comment]
> Design review scheduled for 22nd of October at 10am Pacific Time.

That's at 3-4PM Pacific time.  Sorry for any confustion.

**Comment by mihaimaruseac** [comment]
Notes from the design meeting:

1. Only one question: Base classes that need to implement `__reduce__`:
    * Do we need the `Layer` class to be listed?
      * we're pickling as part of the model
      * Francois: if part of a model, it makes sense to also be available on the layer
1. Francois: High level API: this is very useful for users, good proposal
      * makes sense for models, not so much for optimizers (since state needs to be maintained, we only serialize configs)
      * if all we need is the config, `get_config()` works, if we need state then we need more
         * add some intermediate `SavedModel` serialization
         * or, current RFC only targets model case, is very simple, can be implemented and we'll make a new one for the other case
            * we should not implement at the moment for objects (callbacks, optimizers, metrics, losses with lambdas, etc.) where we only promise to save the config, not the internal state (as this breaks guarantees and API safety; state is usually retrieved via different API calls; results of these can be used to set state too)
1. Callbacks need to be serializable before we can implement this fully
1. RFC only covers built-in objects; in many cases models have custom layers => will the proposal still work?
    * TF function is not something that can be pickled
    * supposed to work with everything that is serializable via SavedModel
      * serialization via `SavedModel` which can save to RAM
      * why not implement this on the `SavedModel`? It is used outside in other packages (i.e., `scikeras`)
1. If we pickle something that we won't support, will we throw an error?
    * with the model, it should work
1. Does this impose new versioning requirements?
    * since it is backed by `SavedModel`, no, just uses `SavedModel`'s versioning
    * `pickle` itself makes no guarantee about versioning, so this is more strict

**Conclusions**:
1. We will restrict scope to `SavedModel` pickling as this solves majority of user requirements / feature requests
1. Proposal accepted

**Review by ematejska — APPROVED** [review]

**Comment by jakirkham** [comment]
Great to see this in! Thanks all! 😄 👏
